### ISSUE TITLE
Mixin Composition Syntax

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#541]
Proposed solution for #540. 

Dart has mixins, and the ability to apply multiple mixins at a time, but no way to combine multiple mixins into a single mixin.

Currently a *mixin* is defined as the difference between a class and its superclass. It is a set of method declarations and a set of interfaces implemented.
Mixin application applies this difference on top of a superclass to create a new class.

It is essentially what:
```dart
class A extends B implements C, D { 
  members
}
```
defines in the `implements C, D { members }` part, as a difference from the `B` superclass, but with a `mixin` declaration you can define this difference independently of the superclass, and then apply the same mixin to a number of superclasses:
```dart
mixin BCD on B implements C, D { members }
class A = B with BCD;
```

You sometimes want to apply more than one mixin:
```dart
mixin IterableMixin<T> implements Iterable<T> { ... }
mixin ListMixin<T> on Iterable<T> implements List<T> { ... }
class IterableBase<T> = Object with IterableMixin<T>;
class ListBase<T> = IterableBase<T> with ListMixin<T>;
class MyList<T> extends ListBase<T> { ... }
class MyList2<T> extends Object with IterableMixin<T>, ListMixin<T> { ... }
```
Here we can combine the iterable and list primitives into on *base class*, but we cannot do the same for mixins. You have to serially apply the mixins to get the same effect.

It would be *nice* and *convenient for users* if we could define a `ListMixin` mixin which *includes* the members of `IterableMixin`, without having to duplicate code.

(The current design actually duplicates code, so `ListMixin` can be applied to something which is not an `Iterable` because it has copied all the members of `IterableMixin`).

## Proposal

Allow `mixin` declarations to extend other mixins.

```dart
mixin ListMixin<T> on Iterable<T> extends IterableMixin<T> {
   ...
}
```
If a mixin with an `extends` clause does not have an `on` clause, it inherits the one of the first mixin it extends. If it has an `on` clause, the `on` clause must be a subtype of the `on` clause of the first mixin it extends.

I say "first" because you can extend more than one mixin:
```dart
mixin Foo on Bar extends Baz, Qux implements Zip, Zap {
  ...
}
```
Here `Bar` must be a subtype of the `on` type of `Baz`, and the type of `Baz` must be a subtype of the `on` type of `Qux`.

The result of this mixin declaration is a *mixin composition* of then three mixins: `Baz` &times; Qux &times; `implements Zip, Zip { ... }` with an `on` type of `Bar`.
Applying a mixin composition is equivalent to applying the three mixins in order.

Inside the mixin composition declaration, members act as if their "superclass" is `Bar with Baz, Qux`. Implementations can do statically resolved super-invocations for concrete members of `Baz` or `Qux`, 

The grammar is modifed as:
```dart
<mixinDeclaration> ::= <metadata> `mixin` <identifier> <typeParamters>? (`on` <type>)? (`extends` <type>(`,` <type>)*)? ((`implements` <type>(`,` <type>)*)? `{` <classMemberDeclaration>* `}`)?
```
A "mixin" is now a difference between a class and a class further up the superclass chain, not necessarily the immediate superclass. A *simple mixin* has one level of class hierarchy, and a *composite* mixin has more.

The body-and-implements parts of a mixin declaration can be omitted. It is a compile-time error to have no body *and* no extends clause. This means that each mixin declaration introduces at least one level of class difference.
```dart
mixin Foo on Bar extends Baz, Qux;
```
This declares the composite of `Baz` and `Qux`, and nothing more, a two-level composite mixin, whereas
```dart
mixin Foo on Bar extends Baz, Qux {}
```
declares a *three*-level composite mixin, with the final level not adding anything.

Semantically, everything is defined in terms of applying a simple mixin to a class. Applying a composite mixin is equivalent to applying the individual levels of the composite in left-to-right (top-to-bottom) order.

**Comment by rrousselGit** [comment]
> Allow mixin declarations to extend other mixins.
> 
>```dart
> mixin ListMixin<T> on Iterable<T> extends IterableMixin<T> {
>    ...
>}
>```

Why not `with` instead of `extends`?

```dart
mixin Foo with Bar, Baz {}
```

Because in a pre `mixin` keyword version, where we used `abstract class` instead, we'd definitely use `with` instead of `extends`:

```dart
abstract class Baz with Foo, Bar {}

class Test with Baz {}
```

**Comment by munificent** [comment]
> Why not `with` instead of `extends`?

+1. I think `with` is the natural keyword to use here. `extends` is something you do with classes.

**Comment by lrhn** [comment]
The `with` keyword means *mixin application*, not just "something you do with mixins". It takes one class and one mixin, and it creates a new class. 

What we are doing here is not that, it is mixin *composition*. It behaves much more like class `extends` (you extend one mixin with more below) than mixin application. It takes one mixin, adds more on top and creates a new mixin. That is "mixin extends" more than it's mixin application.

No choosing `with` was quite deliberate. It might just be that my mental model is not the same as everybody else's, because I make distinction between things that are not that important :smile:.

**Comment by rrousselGit** [comment]
My biggest issue with the `extends` keyword here is that it'd behave like `with`:

- take a list of types, instead of a single type
- accept only mixins

So `extends` may be confusing and make peoples wonder why we can do:

```dart
mixin A {}
mixin B {}
mixin C extends A, B {}
```

but not:

```dart
class A {}
class B {}
class C extends A, B {}
```

nor:

```dart
class A {}
mixin B extends A {}
```

**Comment by lrhn** [comment]
I can see that it behaves *syntactically* like `with`. Maybe that's a hint that it is more like `with` than I think, if I look at it from a different perspective.

If we see `mixin Foo { }` as a "class function" `(S extends Object) => class $ extends S  { ... }` and `with` as such a class-function application, then:
```dart
mixin Foo<T> on Bar<T> with Baz<T>, Qux<T> implements Fifi<T> { }
```
would be the function 
```dart
<T>(S extends Bar<T>) => class $ extends S with Baz<T>, Qux<T> implements Fifi<T> {}
```
It means that the mixin application:
```dart
class C = Bar<int> with Foo<int>;
```
would expand to:
```dart
class C extends Bar<int> with Baz<int>, Qux<int> implements Fifi<int> {}
```

Basically, we copy everything after the `on` type in the mixin declaration into the new class, including the `with` clauses, we are just abstracting over the superclass (which is what mixins are always doing anyway).

So, OK, `with` is acceptable.

**Comment by lrhn** [comment]
The `mixin Foo = Bar with Baz;` feels like it's spuriously making a distinction between the first mixin and a later mixin, but ... again it might make sense to users.

If there is an `on` clause, I think it might read better as `mixin Foo on Qux with Bar, Baz;`. Not a lot, though.

**Comment by icatalud** [comment]
I remember spending hours trying to make mixin composition work. I was so dissapointed that I couldn't compose mixins. I think it should be explicit in language Tour mixin definition that this is not possible yet (this information is only available by digging in the specification), because they "feel" like classes and it might be happening that many people intuitively are attempting to compose mixins (trying all possible combinations of extends, with, etc). I even remember that I experimented something like `abstract class A = Object with Mix1, Mix2` then I could do `class B with A` and it would pass the static type checks (program would throw exception).

I also prefer `with M1, M2` over `extends`, but I agree it's a bit weird, how about plain equality (I believe I saw equality proposed on the Mixin specification), like:

`mixin D = A, B, C`

Where A, B and C are mixins. Perhaps it is less intuitive, because people will naturally attempt to use them like classes with `extend` or `with`, but defining classes with equality is kinda like a "Dart thing" (I saw that pattern in Dart for the first time and I like it).

**Comment by arcra** [comment]
Just as @icatalud described, I just spent a good amount of time trying to learn how to use Mixins.

My goal was to make something like this work:

```
mixin Incrementer {
  int value = 1;

  void increment() {
    value++;
  }
}

mixin StepIncrementer extends Incrementer {
  int get step;

  @override
  void increment() {
    value += step;
  }
}

// Implementation does not need to be repeated.
mixin OneStepIncrementerMixin extends StepIncrementer {
  final int step = 1;
}

mixin TwoStepIncrementerMixin extends StepIncrementer {
  final int step = 2;
}

// A implements all: TwoStepIncrementer, StepIncrementer and Incrementer.
class A with TwoStepIncrementer { }

var a = A()..increment();
print(a.value); // Prints 3;
```

I understand I might be doing something wrong. I don't really understand the technicalities of implementing this feature, but that's precisely why I generally agree with this statement:

> I think it should be explicit in language Tour mixin definition that this is not possible yet (this information is only available by digging in the specification), because they "feel" like classes and it might be happening that many people intuitively are attempting to compose mixins (trying all possible combinations of extends, with, etc).

As a language user, it took me longer than I would have liked to understand that it couldn't be done (and to find clarifying information like the one in this issue).

**Comment by lrhn** [comment]
I've come back to thinking about this, and with the new `mixin class` syntax, I'm more positive towards using `with` to compose mixins.

That is:
```dart
mixin M with M1, M2 { ... }
```
could define a named mixin `M` which is the composition of `M1`, `M2` and `{...}`. You can apply it just by doing `with M`, which would apply `M1`, `M2` and `{...}` in that order. (There will be three superclasses in the superclass chain, otherwise any `super.` invocations wouldn't work correctly.)

You may also define `mixin M with M1, M2;` without a body, to just compose `M1` with `M2`.

For mixin classes, you'd be allowed to do `mixin class ListMixin with IterableMixin { .... }` and be allowed to derive the mixin composition `IterableMixin & {....}` from the declaration, and also declare `mixin class C with M1, M2;` with no body if you really want to.

(We can even consider re-allowing `mixin class extends S with M1, M2;` to derive a mixin that must be applied to `S`. Now that the class is explicitly marked `mixin`, we know that it's intended and that we should treat `super` invocations specially. Probably not a strong need for that, but might feel a little more orthogonal.)
A named composite mixin refers to all the mixins applied in that named declaration, and they can be applied together,
so `class MySomething with ListMixin<E>` will apply both `IterableMixin` and `ListMixin`'s `{...}`.

**Comment by TekExplorer** [comment]
> I've come back to thinking about this, and with the new `mixin class` syntax, I'm more positive towards using `with` to compose mixins.
> 
> That is:
> 
> mixin M with M1, M2 { ... }
> could define a named mixin `M` which is the composition of `M1`, `M2` and `{...}`. You can apply it just by doing `with M`, which would apply `M1`, `M2` and `{...}` in that order. (There will be three superclasses in the superclass chain, otherwise any `super.` invocations wouldn't work correctly.)
> 
> You may also define `mixin M with M1, M2;` without a body, to just compose `M1` with `M2`.
> 
> For mixin classes, you'd be allowed to do `mixin class ListMixin with IterableMixin { .... }` and be allowed to derive the mixin composition `IterableMixin & {....}` from the declaration, and also declare `mixin class C with M1, M2;` with no body if you really want to.

I really like that actually.
I think its a really important syntax to allow, as right now mixins are almost useless besides super simple usecases, when it _should_ be way more powerful.
I'm _constantly_ frustrated that I can't really use mixins because it ends up ballooning, or I end up needing to make a superclass anyway, which then blocks future use of superclass, or now I need to implement when there is no actual need to. it sucks.

> (We can even consider re-allowing `mixin class extends S with M1, M2;` to derive a mixin that must be applied to `S`. Now that the class is explicitly marked `mixin`, we know that it's intended and that we should treat `super` invocations specially. Probably not a strong need for that, but might feel a little more orthogonal.) A named composite mixin refers to all the mixins applied in that named declaration, and they can be applied together, so `class MySomething with ListMixin<E>` will apply both `IterableMixin` and `ListMixin`'s `{...}`.

that's neat, though I figure its something that can come after.

**Comment by TekExplorer** [comment]
Surprisingly, we can actually currently do
```dart
abstract mixin class X = Object with M;

// works fine!
// not that useful though...
// good for extensions maybe?
class Foo with X {}
```
and that works fine, which is _curious_ because this does _not_ work
```dart
// so close
// the class X cannot be a mixin because it extends a class other than `Object`
// which is actually kind of a misleading error here
abstract mixin class X = Object with M, M2;

// of course giving it a second one wouldn't work either.
// the class X cannot be a mixin because it extends a class other than `Object`
abstract mixin class X with M {}
```
odd.

**Comment by lrhn** [comment]
The error is correct. A mixin class declaration must have a `Object` as superclass. That's actual superclass, the superclass of the class being declared, not _declared superclass_, which is what's in the `extends` clause. 

The class `X` of
```dart
abstract mixin class X = Object with M, M2;
```
Has `Object with M` as actual superclass.

**Comment by TekExplorer** [comment]
> The error is correct. A mixin class declaration must have a `Object` as superclass. That's actual superclass, the superclass of the class being declared, not _declared superclass_, which is what's in the `extends` clause.
> 
> The class `X` of
> 
> abstract mixin class X = Object with M, M2;
> Has `Object with M` as actual superclass.

I understand that - the errors i expected. its the `abstract mixin class X = Object with M;` that i find odd. i wonder why that works, relatively useless as it is.
